docs/proposals/202004_embedd_cortex_frontend.md

```diff
@@ -0,0 +1,119 @@
+We even started discussion to move it to separate repo, but there was no motivation towards this, for a good reason.
+At the end we were advertising to use cortex query frontend on production on top of Thanos and this works considerably well, with some
+problems on edge cases and for downsampled data as mentioned [here]()
+
+However, we realized recently that asking users to install suddenly Cortex component on top of Thanos system is extremely confusing:
+
+* Cortex has totally different way of configuring services. It requires to decide what module you have in single YAML file. Thanos in opposite
+have flags and different subcommand for each component.
+* Cortex has bit different way of configuring memcached, which is inconsistent with what we have in Thanos Stote Gateway.
+* There are many Cortex specific configuration items which can confuse Thanos user and increase complexity overall (e.g. )
+* We have many ideas how to improve Cortex Query Frontend on top of Thanos, but adding Thanos specific configuration options will increase
+complexity on Cortex side as well.
+* Cortex has no good example or tutorial on how to use frontend either. We have only [Observatorium example](https://github.com/observatorium/configuration/blob/master/environments/openshift/manifests/observatorium-template.yaml#L515).
+
+All of this were causing confusion and questions like [this](https://cloud-native.slack.com/archives/CK5RSSC10/p1586504362400300?thread_ts=1586492170.387900&cid=CK5RSSC10).
+
+At the end we decided with Thanos and Cortex maintainers that the ultimately it would be awesome to create a new Thanos subcommand called `frontend`.
+
+## Use Cases
+
+* User can cache responses for query range
+* User can use the same configuration patterns as rest of Thanos components.
+* Thanos Developer does not need to write and maintain custom response cache logic.
+
+## Goals of this design
+
+* Enable response caching that will easy to use for Thanos users.
+* Keep it extensible and scalable for future improvements like advanced query planning, queuing, rate limiting etc.
+* Reuse as much as possible between projects, contribute.
+
+## No Goals
+
+* Create Thanos specific response caching from scratch.
+
+## Proposal
+
+The idea is to create `thanos frontend` component that allows specifying following options:
+
+* `--query-range.split-interval`, `time.Duration`
+* `--query-range.max-retries-per-request`, `int`, default = `5`
+* `--query-range.disable-step-align`, `bool`
+* `--query-range.response-cache-ttl` `time.Duration` default = `1m`
+* `--query-range.response-cache-config(-file)` `pathorcontent` + [CacheConfig](https://github.com/thanos-io/thanos/blob/55cb8ca38b3539381dc6a781e637df15c694e50a/pkg/store/cache/factory.go#L32)
+
+We plan to have in-mem, fifo and memcached support for now. Cache config will be exactly the same as the one used for Store Gateway.
+
+### Open Questions
+
+* Is `thanos frontend` a valid name? I feel it's short and verbose enough.
```

Comment:
I have a feeling that people new to the project are going to be confused about this. Why not `query-cache`?

Reply:
Hm, I disagree here. While we can discuss better ideas `query-cache` is particularly not perfect one. It clash with `query` so again huge confusion. Additonally this component introduces much more than cache. I would rather call it.. `lb`? or `loadbalance`? It's just a frontend proxy with features.

Maybe:
* `loadbalance`
* `lb`
* `ingress` 
(:

Reply:
Also one-worder would be nice

Reply:
My take:
- Frontend can be confused with "UI frontend" (as of Today, I wouldn't call it frontend in Cortex)
- `query-cache` looks reductive. Even if now the primary use case could be caching, we may end up putting in this component other read path ingress logic

> loadbalance
> lb

I think it's reductive too (because doesn't just do load balancing).

> ingress

Better option to me.

Reply:
> Also one-worder would be nice

Agreed, but if we can't find something better, I wouldn't mind it too much.

> query-cache looks reductive. Even if now the primary use case could be caching, we may end up putting in this component other read path ingress logic

Agreed.

Since a very common target platform is Kubernetes, I'm not too sure about ingress either. Maybe Cortex's naming of `query-frontend` isn't all that bad?

Reply:
`ingress` looks good, just curious if it's not mistaken / confused by Kubernetes ingress ):

Reply:
What about... `edge`?

Reply:
`planner`?

Reply:
`pre-query`

Reply:
naming is hard ):

Reply:
Asked on slack for some opinions, let's look deeper. I don't like query-frontend. Still exacty same confusion with both UI and queried.

Reply:
> Since a very common target platform is Kubernetes, I'm not too sure about ingress either. Maybe Cortex's naming of `query-frontend` isn't all that bad?

I like `query-frontend`. It seems descriptive and it matches what Cortex are doing.

Reply:
`proxy`?

Reply:
Proxy might be very confusing as Querier is literally a proxy, nothing more ;p

Reply:
Looks like majority on slack like consistency :heart:  so `query-frontend`: https://cloud-native.slack.com/archives/CK5RSSC10/p1586942580182000

Of course @brancz was right again (:

Reply:
I propose to call in code `frontend` but call in docs everywhere query-frontend. Same as we do with store. (:

Reply:
thoughts?

Reply:
> I propose to call in code frontend

I'm OK calling it frontend in the code, but a bit dubious about calling the cmd `thanos frontend` (I understand this is the same trade-off picked for `store`, but at the beginning I found it confusing too). I would personally pick clarity on the cmd, so `thanos query-frontend` (given this is the naming we're going to use in the doc), at the cost of having to type 6 more characters 😉

Reply:
I think the whole discussion here was about the command, so I still think the command should be query-frontend.